src/ray/common/ray_syncer/ray_syncer-inl.h

```diff
@@ -0,0 +1,201 @@
+};
+
+class NodeSyncConnection {
+ public:
+  NodeSyncConnection(RaySyncer &instance,
+                     instrumented_io_context &io_context,
+                     std::string node_id);
+
+  /// Push a message to the sending queue to be sent later. Some message
+  /// might be dropped if the module think the target node has already got the
+  /// information. Usually it'll happen when the message has the same node id
+  /// as the target or the message is sent from this node.
+  ///
+  /// \param message The message to be sent.
+  ///
+  /// \return true if push to queue successfully.
+  bool PushToSendingQueue(std::shared_ptr<const RaySyncMessage> message);
+
+  /// Send the message queued.
+  virtual void DoSend() = 0;
+
+  virtual ~NodeSyncConnection() {}
+
+  /// Return the node id of this sync context.
+  const std::string &GetNodeId() const { return node_id_; }
+
+  /// Handle the udpates sent from this node.
+  ///
+  /// \param messages The message received.
+  void ReceiveUpdate(RaySyncMessages messages);
+
+ protected:
+  FRIEND_TEST(RaySyncerTest, NodeSyncConnection);
+
+  std::array<int64_t, kComponentArraySize> &GetNodeComponentVersions(
+      const std::string &node_id);
+
+  RaySyncer &instance_;
+  instrumented_io_context &io_context_;
+  std::string node_id_;
+
+  struct _MessageHash {
+    std::size_t operator()(
+        const std::shared_ptr<const RaySyncMessage> &m) const noexcept {
+      std::size_t seed = 0;
+      boost::hash_combine(seed, m->node_id());
+      boost::hash_combine(seed, m->component_id());
+      return seed;
+    }
+  };
+
+  struct _MessageEq {
+    bool operator()(const std::shared_ptr<const RaySyncMessage> &lhs,
+                    const std::shared_ptr<const RaySyncMessage> &rhs) const noexcept {
+      if (lhs == rhs) {
+        return true;
+      }
+      if (lhs == nullptr || rhs == nullptr) {
+        return false;
+      }
+      // We don't check the version here since we want the old version to be deleted.
+      return lhs->node_id() == rhs->node_id() &&
+             lhs->component_id() == rhs->component_id();
+    }
+  };
+
+  absl::flat_hash_set<std::shared_ptr<const RaySyncMessage>, _MessageHash, _MessageEq>
+      sending_queue_;
+  // Keep track of the versions of components in this node.
+  absl::flat_hash_map<std::string, std::array<int64_t, kComponentArraySize>>
+      node_versions_;
+};
+
+class ServerSyncConnection : public NodeSyncConnection {
+ public:
+  ServerSyncConnection(RaySyncer &instance,
+                       instrumented_io_context &io_context,
+                       const std::string &node_id);
+
+  ~ServerSyncConnection() override;
+
+  void HandleLongPollingRequest(grpc::ServerUnaryReactor *reactor,
+                                RaySyncMessages *response);
+
+ protected:
+  /// Send the message from the pending queue to the target node.
+  /// It'll send nothing unless there is a request from the remote node
+  /// for the sending request.
```

Comment:
What does this sentence mean?

Reply:
Oh, this is about long-polling. It means unless the remote node request, it won't send the data.